File: Algorithms/0983_Minimum_Cost_For_Tickets/Java/MinimumCostForTicketsSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        if (days.length == 0) return 0;
        if (days.length == 1) return costs[0];

        int[] dp = new int[days[days.length - 1] + 1];
        Arrays.fill(dp, 0);
        int index = 0;
        for (int i = 1; i < dp.length; i++) {
            if (index < days.length && days[index] == i) {
                int dayPass = dp[i - 1] + costs[0];
                int weekPass = dp[Math.max(0, i - 7)] + costs[1];
                int monthPass = dp[Math.max(0, i - 30)] + costs[2];
                dp[i] = Math.min(dayPass, Math.min(weekPass, monthPass));
                index++;
            } else {
                dp[i] = dp[i - 1];
            }
        }
        return dp[dp.length - 1];
    }
}

File: Algorithms/0983_Minimum_Cost_For_Tickets/Python/Minimum_Cost_For_Tickets_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def mincostTickets(self, days, costs) -> int:
        num_of_days = len(days)
        if num_of_days == 0: return 0
        if num_of_days == 1: return costs[0]

        dp = [0 for _ in range(days[-1] + 1)]
        for i in range(1, len(dp)):
            if i not in days:
                dp[i] = dp[i - 1]
            else:
                day_pass = dp[i - 1] + costs[0]
                week_pass = dp[max(0, i - 7)] + costs[1]
                month_pass = dp[max(0, i - 30)] + costs[2]
                dp[i] = min(day_pass, week_pass, month_pass)

        return dp[-1]




